Neopixel background LED option

Bob-the-Kuhn opened the pull request:
This PR adds the ability to designate one of the neopixel LEDs as the background LED and to set it's color.  The color of this LED is set after reset and can't be changed by M150 or by the printer_event_leds code.

After reset, the LEDs will go through the NEOPIXEL_STARTUP_TEST (if enabled) , set all the non-background LED colors and then set the background LED color.

This is mostly so the FYSETC_MINI_12864_2_1 can have a constant color backlight.

---

Configuration.h - added defines BACKGROUND_NEOPIXEL_LED and
BACKGROUND_NEOPIXEL_LED_COLOR

leds.cpp - set_color - skip setting LED if index equals background LED number

neopixel.cpp - neopixel_color- skip setting LED if index equals background LED number

neopixel.cpp - setup_neopixel - added setting of background LED color

Bob-the-Kuhn committed:
add backiground noepixel LED function

Configuration.h - added defines BACKGROUND_NEOPIXEL_LED and
BACKGROUND_NEOPIXEL_LED_COLOR

leds.cpp - set_color - skip setting LED if index equals background LED number

neopixel.cpp - neopixel_color- skip setting LED if index equals background LED number

neopixel.cpp - setup_neopixel - added setting of background LED color

Bob-the-Kuhn committed:
add backiground noepixel LED function

Configuration.h - added defines BACKGROUND_NEOPIXEL_LED and
BACKGROUND_NEOPIXEL_LED_COLOR

leds.cpp - set_color - skip setting LED if index equals background LED number

neopixel.cpp - neopixel_color- skip setting LED if index equals background LED number

neopixel.cpp - setup_neopixel - added setting of background LED color

Bob-the-Kuhn committed:
neopixel.cpp update for compatibility with LPC1768 neopixel library

Added a set_color routine specifically for the startup test.  This allows all LEDs to be tested during the startup while maintaining the independence of the background LED during normal operations.

Bob-the-Kuhn committed:
remove bad commit and its revert

Added a set_color routine specifically for the startup test.  This allows all LEDs to be tested during the startup while maintaining the independence of the background LED during normal operations.

Bob-the-Kuhn commented:
Added a set_color specifically for the neopixel startup test.   The new routine sets all LEDs.

Testing the code with the LPC1768 neopixel library showed that the startup test was depending on a bug in the current library.

Bob-the-Kuhn committed:
Merge branch 'neopixel-background-color' of https://github.com/Bob-the-Kuhn/Marlin-Bob-2 into neopixel-background-color

Bob-the-Kuhn committed:
yet another attempt to get rif of that line

Bob-the-Kuhn committed:
formatting cleanup in example Configuration.h files

Bob-the-Kuhn committed:
formatting cleanup on neopixel.cpp

p3p commented:
> Testing the code with the LPC1768 neopixel library showed that the startup test was depending on a bug in the current library.

Bug in the LPC lib? if its out of date I can rebase

Bob-the-Kuhn commented:
No - the bug was in the library used by everything other than the LPC1768.  

I had noticed that the startup test was also setting the background color despite my code.  When I went to the LPC1768's library the system worked as I expected.  I then added a set color routine specific to the startup test so that the background color was guaranteed to be set during the startup test.

thinkyhead commented on Marlin/src/feature/leds/neopixel.cpp:
```diff
@@ -50,14 +68,18 @@ void setup_neopixel() {
   #endif
 
+  #ifdef BACKGROUND_NEOPIXEL_LED
+    set_neopixel_color_background();
```
What does this do when `BACKGROUND_NEOPIXEL_LED` is zero?

thinkyhead replied:
Oh, I see. I’ll change the description to “index of LED” rather than “number of LED” so fools like me and non-English-speakers won’t interpret it as “quantity of LEDs” that don’t change.

Bob-the-Kuhn commented:
How about?
**NOW:**
```cpp
#define BACKGROUND_NEOPIXEL_LED 0  // number of LED in chain that never changes color
```
**TO:**
```cpp
#define BACKGROUND_NEOPIXEL_LED_INDEX 0  // 0, 1, 2 ..  index of LED in chain that never changes color
```

Bob-the-Kuhn committed:
change name of variable to make its role clearer

NOW:
#define BACKGROUND_NEOPIXEL_LED 0 // number of LED in chain that never changes color
TO:
#define BACKGROUND_NEOPIXEL_LED_INDEX 0 // 0, 1, 2 .. index of LED in chain that never changes color

thinkyhead committed:
Use names with a consistent style

Bob-the-Kuhn commented:
I'm OK with Scott's changes.

I'm OK with merging this PR.

Christophegriffo commented:
salut je reçoit un message erreur quand j'utilise #define NEOPIXEL_BKGD_LED_INDEX 0 
![Capture](https://user-images.githubusercontent.com/53877827/64242589-1a545a80-cf06-11e9-89a4-e4a92dca701d.PNG)
je pense que c'est casser. pouvez regarder d'ou viens le problème 
merci

Patag commented:
Comment line 73 of Neopixel.h
```
  #ifdef NEOPIXEL_BKGD_LED_INDEX
    //static void set_pixel_color(const uint16_t n, const uint32_t c);
  #endif
```

Christophegriffo commented:
@Patag ca fonctionne merci bien. a tout hasard tu sais pourquoi le rouge s'allume alors que j'ai mis R0 G0 B255 W0. j'ai bien mis neoRGB
